kernel/src/table_changes/scan_file.rs

```diff
@@ -0,0 +1,405 @@
+            let timestamp = getters[16].get(row_index, "scanFile.timestamp")?;
+            let commit_version: i64 = getters[17].get(row_index, "scanFile.commit_version")?;
+
+            let (scan_type, path, deletion_vector, partition_values) =
+                if let Some(path) = getters[0].get_opt(row_index, "scanFile.add.path")? {
+                    let scan_type = CDFScanFileType::Add;
+                    let deletion_vector = visit_deletion_vector_at(row_index, &getters[1..=5])?;
+                    let partition_values = getters[6]
+                        .get(row_index, "scanFile.add.fileConstantValues.partitionValues")?;
+                    (scan_type, path, deletion_vector, partition_values)
+                } else if let Some(path) = getters[7].get_opt(row_index, "scanFile.remove.path")? {
+                    let scan_type = CDFScanFileType::Remove;
+                    let deletion_vector = visit_deletion_vector_at(row_index, &getters[8..=12])?;
+                    let partition_values = getters[13].get(
+                        row_index,
+                        "scanFile.remove.fileConstantValues.partitionValues",
+                    )?;
+                    (scan_type, path, deletion_vector, partition_values)
+                } else if let Some(path) = getters[14].get_opt(row_index, "scanFile.cdc.path")? {
+                    let scan_type = CDFScanFileType::Cdc;
+                    let partition_values = getters[15]
+                        .get(row_index, "scanFile.cdc.fileConstantValues.partitionValues")?;
+                    (scan_type, path, None, partition_values)
+                } else {
+                    continue;
+                };
+            let dv_info = DvInfo { deletion_vector };
+            let scan_file = CDFScanFile {
+                scan_type,
+                path,
+                dv_info,
+                partition_values,
+                commit_version,
+                timestamp,
+            };
+            (self.callback)(&mut self.context, scan_file)
+        }
+        Ok(())
+    }
+
+    fn selected_column_names_and_types(&self) -> (&'static [ColumnName], &'static [DataType]) {
+        static NAMES_AND_TYPES: LazyLock<ColumnNamesAndTypes> =
+            LazyLock::new(|| cdf_scan_row_schema().leaves(None));
+        NAMES_AND_TYPES.as_ref()
+    }
+}
+
+/// Get the schema that scan rows (from [`TableChanges::scan_data`]) will be returned with.
+pub(crate) fn cdf_scan_row_schema() -> SchemaRef {
+    static CDF_SCAN_ROW_SCHEMA: LazyLock<Arc<StructType>> = LazyLock::new(|| {
+        let deletion_vector = StructType::new([
+            StructField::new("storageType", DataType::STRING, true),
+            StructField::new("pathOrInlineDv", DataType::STRING, true),
+            StructField::new("offset", DataType::INTEGER, true),
+            StructField::new("sizeInBytes", DataType::INTEGER, true),
+            StructField::new("cardinality", DataType::LONG, true),
+        ]);
```

Comment:
> should we just use DeletionVectorDescriptor::to_schema instead

I'd like to, but we hard code the number of fields in the visitor getters. So any change to the DV schema would break the visitor.

The schema stuff makes sense! I'll keep the top level fields nullable and patch up the nested ones.

Reply:
Also given that partitionValues is technically nullable for removes, I'm constructing a default map if I don't find anything. Is this necessary/sensible?
```rust
let partition_values = partition_values.unwrap_or_else(Default::default);
```

Reply:
> Found unmasked nulls for non-nullable StructArray field \"path\"

That error indicates that `add` was non-null but `path` was null, which is definitely not a good thing. Do you have a backtrace? I wonder if this tickled another latent bug in our arrow data code, similar to the one that @nicklan hit a while back?

Reply:
(also just in case -- did you verify that the underlying data are actually correct? we have tripped over invalid data a couple times in the past)

Reply:
Hmm I'll try to track this down and report back 🫡

Reply:
Okay I've narrowed down the point in the code where we fail for this. It's actually in the expression transformation, and not the visitor!

The specific line is [here](https://github.com/delta-io/delta-kernel-rs/blob/main/kernel/src/engine/arrow_expression.rs#L411). I don't see anything that handles nested nullability, but maybe arrow is supposed to do that on its own?

I think I've got a reasonably MRE. I'll post an issue.

Reply:
I submitted an issue. Would love for you to sanity check it @scovich: 
https://github.com/delta-io/delta-kernel-rs/issues/578

Reply:
I'll punt on this for now and keep everything nullable. This is what we do in the normal read path, so I think it should be okay for now.

Reply:
attn @nick